litex/soc/interconnect/avalon.py

```diff
@@ -9,6 +9,156 @@
 from migen import *
 
 from litex.soc.interconnect import stream
+from . import wishbone
+
+_layout = [
+    ("address",          "adr_width", DIR_M_TO_S),
+    ("writedata",       "data_width", DIR_M_TO_S),
+    ("readdata",        "data_width", DIR_S_TO_M),
+    ("readdatavalid",              1, DIR_S_TO_M),
+    ("byteenable",       "sel_width", DIR_M_TO_S),
+    ("read",                       1, DIR_M_TO_S),
+    ("write",                      1, DIR_M_TO_S),
+    ("waitrequest",                1, DIR_S_TO_M),
+    ("burstbegin",                 1, DIR_M_TO_S), # this is optional
+    ("burstcount",                 8, DIR_M_TO_S),
+    ("chipselect",                 1, DIR_M_TO_S), # this is optional
+]
+
+class MMInterface(Record):
+    def __init__(self, data_width=32, adr_width=30, bursting=False, **kwargs):
+        self.data_width = data_width
+        self.bursting   = bursting
+        if kwargs.get("adr_width", False):
+            adr_width = kwargs["adr_width"] - int(log2(data_width//8))
+        self.adr_width = adr_width
+        Record.__init__(self, set_layout_parameters(_layout,
+            adr_width  = adr_width,
+            data_width = data_width,
+            sel_width  = data_width//8))
+        self.address.reset_less    = True
+        self.writedata.reset_less  = True
+        self.readdata.reset_less   = True
+        self.byteenable.reset_less = True
+
+    @staticmethod
+    def like(other):
+        return MMInterface(len(other.writedata))
+
+    def get_ios(self, bus_name="avl"):
+        subsignals = []
+        for name, width, direction in self.layout:
+            subsignals.append(Subsignal(name, Pins(width)))
+        ios = [(bus_name , 0) + tuple(subsignals)]
+        return ios
+
+    def connect_to_pads(self, pads, mode="master"):
+        assert mode in ["slave", "master"]
+        r = []
+        for name, width, direction in self.layout:
+            sig  = getattr(self, name)
+            pad  = getattr(pads, name)
+            if mode == "master":
+                if direction == DIR_M_TO_S:
+                    r.append(pad.eq(sig))
+                else:
+                    r.append(sig.eq(pad))
+            else:
+                if direction == DIR_S_TO_M:
+                    r.append(pad.eq(sig))
+                else:
+                    r.append(sig.eq(pad))
+        return r
+
+    def bus_read(self, address, byteenable=None, burstcount=None, chipselect=None):
```

Comment:
`bus_read `-> `read` for consistency with Wishbone.

Reply:
That does not work, because avalon has  a read signal, which gets in the way.

Reply:
Do you have an idea for a more elegant workaround?

Reply:
Ah thanks, I see. We can keep `bus_read/write` for now then. I'll think about it. Maybe we should make this separate (also for Wishbone) and make it clear it's for simulation.